element_array_ephys/export/nwb/nwb.py

```diff
@@ -0,0 +1,600 @@
+    if all(x == gains[0] for x in gains):
+        return dict(conversion=1e-6 * gains[0], channel_conversion=None)
+    return dict(conversion=1e-6, channel_conversion=gains)
+
+
+def add_ephys_recording_to_nwb(
+    session_key: dict, ephys_root_data_dir,
+    nwbfile: pynwb.NWBFile, end_frame: int = None):
+    """Read voltage data directly from source files and iteratively transfer them to the NWB file. Automatically
+    applies lossless compression to the data, to the final file might be smaller than the original, but there is no
+    data loss. Currently supports neuropixel and openephys, and relies on SpikeInterface to read the data.
+
+    source data -> acquisition["ElectricalSeries"]
+
+    Parameters
+    ----------
+    session_key: dict
+    nwbfile: NWBFile
+    end_frame: int, optional
+        Used for small test conversions
+    """
+
+    if nwbfile.electrodes is None:
+        add_electrodes_to_nwb(session_key, nwbfile)
+
+    mapping = get_electrodes_mapping(nwbfile.electrodes)
+
+    for ephys_recording_record in (ephys.EphysRecording & session_key).fetch(
+        as_dict=True
+    ):
+        probe_id = (ephys.ProbeInsertion() & ephys_recording_record).fetch1("probe")
+
+        relative_path = (
+            ephys.EphysRecording.EphysFile & ephys_recording_record
+        ).fetch1("file_path")
+        relative_path = relative_path.replace("\\","/")
+        file_path = find_full_path(ephys_root_data_dir, relative_path)
+
+        if ephys_recording_record["acq_software"] == "SpikeGLX":
+            extractor = extractors.read_spikeglx(os.path.split(file_path)[0], "imec.ap")
+        elif ephys_recording_record["acq_software"] == "OpenEphys":
+            extractor = extractors.read_openephys(file_path, stream_id="0")
+        else:
+            raise ValueError(
+                f"unsupported acq_software type: {ephys_recording_record['acq_software']}"
+            )
+
+        conversion_kwargs = gains_helper(extractor.get_channel_gains())
+
+        if end_frame is not None:
+            extractor = extractor.frame_slice(0, end_frame)
+
+        recording_channels_by_id = (
+            probe.ElectrodeConfig.Electrode() & ephys_recording_record
+        ).fetch("electrode")
+
+        nwbfile.add_acquisition(
+            pynwb.ecephys.ElectricalSeries(
+                name=f"ElectricalSeries{ephys_recording_record['insertion_number']}",
+                description=str(ephys_recording_record),
+                data=SpikeInterfaceRecordingDataChunkIterator(extractor),
+                rate=ephys_recording_record["sampling_rate"],
+                starting_time=(
+                    ephys_recording_record["recording_datetime"]
+                    - ephys_recording_record["session_datetime"]
+                ).total_seconds(),
+                electrodes=nwbfile.create_electrode_table_region(
+                    region=[mapping[(probe_id, x)] for x in recording_channels_by_id],
```

Comment:
Should this be `probe_id` or `shank_id`?  See note above in `get_electrodes_mapping` function.

Reply:
This thread is resolved.  Since the `electrode` attribute in the `probe.ProbeType.Electrode` table is unique among all shanks in a probe, the mapping is correct as is.